kbkdf/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+//! for Key Derivation Using Pseudorandom Functions][1]
+//!
+//! [1]: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-108r1.pdf
+
+#![no_std]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+use core::{fmt, marker::PhantomData, num::Wrapping, ops::Mul};
+use digest::{
+    consts::{U32, U8},
+    crypto_common::KeySizeUser,
+    generic_array::{typenum::Unsigned, ArrayLength, GenericArray},
+    typenum::op,
+    KeyInit, Mac,
+};
+use divrem::DivCeil;
+
+mod sealed;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    // TODO(baloo): we can probably move that to a compilation error via use of typenum
+    InvalidRequestSize,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidRequestSize => write!(
+                f,
+                "Request output size is too large for the value of R specified"
+            ),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+mod std_error {
+    use super::Error;
+    use std::error;
+
+    impl error::Error for Error {}
+}
+
+// Helper structure along with [`KbkdfUser`] to compute values of L and H.
+struct KbkdfCore<OutputLen, PrfOutputLen> {
+    _marker: PhantomData<(OutputLen, PrfOutputLen)>,
+}
+
+trait KbkdfUser {
+    // L - An integer specifying the requested length (in bits) of the derived keying material
+    // KOUT. L is represented as a bit string when it is an input to a key-derivation function. The
+    // length of the bit string is specified by the encoding method for the input data.
+    type L;
+
+    // h - An integer that indicates the length (in bits) of the output of a single invocation of the
+    // PRF.
+    type H;
+}
+
+impl<OutputLen, PrfOutputLen> KbkdfUser for KbkdfCore<OutputLen, PrfOutputLen>
+where
+    OutputLen: ArrayLength<u8> + Mul<U8>,
+    <OutputLen as Mul<U8>>::Output: Unsigned,
+    PrfOutputLen: ArrayLength<u8> + Mul<U8>,
+    <PrfOutputLen as Mul<U8>>::Output: Unsigned,
+{
+    type L = op!(OutputLen * U8);
+    type H = op!(PrfOutputLen * U8);
+}
+
+/// [`Kbkdf`] is a trait representing a mode of KBKDF.
+/// It takes multiple arguments:
+///  - Prf - the Pseudorandom Function to derive keys from
+///  - K - the expected output length of the newly derived key
+///  - R - An integer (1 <= r <= 32) that indicates the length of the binary encoding of the counter i
+///        as an integer in the interval [1, 2r − 1].
+pub trait Kbkdf<Prf, K, R: sealed::R>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArrayLength<u8> + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArrayLength<u8> + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+{
+    /// Derives `key` from `kin` and other parameters.
+    fn derive(
+        &self,
+        kin: &GenericArray<u8, Prf::KeySize>,
```

Comment:
This is incorrect and doesn't work with hmac.
I think I'll need for it to pass in the `Prf` and reset it.